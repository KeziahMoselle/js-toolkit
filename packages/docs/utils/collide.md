---
sidebar: auto
---

# Collide

Collides utils is used to test if two DOM elements are overlapping.

Inspired by http://www.jeffreythompson.org/collision-detection/

## `collidePointRect`

Test if a point collides with a rectangle.

**Parameters**

- `point` (`Point`)
- `rect` (`Rect`)

[Source](https://github.com/studiometa/js-toolkit/blob/master/src/utils/collide/collidePointRect.js)

**Usage**

```js
import { collidePointRect } from '@studiometa/js-toolkit/utils/collide';

const point = {
  x: 0,
  y: 0,
};

const rect = {
  x: 0,
  y: 0,
  width: 20,
  height: 20,
};

collidePointRect(point, rect); // true
```

## `collideRectRect`

Test if a rectangle collides with another rectangle.

**Parameters**

- `rect1` (`Rect`)
- `rect2` (`Rect`)

[Source](https://github.com/studiometa/js-toolkit/blob/master/src/utils/collide/collideRectRect.js)

**Usage**

```js
import { collideRectRect } from '@studiometa/js-toolkit/utils/collide';

const rect1 = {
  x: 0,
  y: 0,
  width: 20,
  height: 20,
};

const rect2 = {
  x: 0,
  y: 0,
  width: 20,
  height: 20,
};

collideRectRect(rect1, rect2); // true
```

## `boundingRectToCircle`

Convert clientRect to a formatted circle object.

⚠️ It doesn't work if the DOM element is not a perfect square, but it can be forced with the second parameter.

**Parameters**

- `domRect` (`Partial<DOMRect>`)
- `force` (`Boolean`) : `false` by default

[Source](https://github.com/studiometa/js-toolkit/blob/master/src/utils/collide/boundingRectToCircle.js)

**Usage**

```js
import { boundingRectToCircle } from '@studiometa/js-toolkit/utils/collide';

const clientRect1 = { x: 0, y: 0, width: 100, height: 100 };
boundingRectToCircle(clientRect1); // { x: 50, y: 50, radius: 50 }

const clientRect2 = element.getBoundingClientRect();
boundingRectToCircle(clientRect2);
```

## `collideCircleRect`

Test if a circle collides with a rectangle.

**Parameters**

- `circle` (`Circle`)
- `rect` (`Rect`)

[Source](https://github.com/studiometa/js-toolkit/blob/master/src/utils/collide/collideCircleRect.js)

**Usage**

```js
import { collideCircleRect } from '@studiometa/js-toolkit/utils/collide';

const circle = {
  x: 40,
  y: 40,
  radius: 40,
};

const rect = {
  x: 40,
  y: 40,
  width: 60,
  height: 60,
};

collideCircleRect(circle, rect); // true
```

## `collideCircleCircle`

Test if a circle collides with another.

**Parameters**

- `circle1` (`Circle`)
- `circle2` (`Circle`)

[Source](https://github.com/studiometa/js-toolkit/blob/master/src/utils/collide/collideCircleCircle.js)

**Usage**

```js
import { collideCircleCircle } from '@studiometa/js-toolkit/utils/collide';

const circle1 = {
  x: 40,
  y: 40,
  radius: 40,
};

const circle2 = {
  x: 100,
  y: 100,
  radius: 60,
};

collideCircleCircle(circle1, circle2); // true
```

## `collidePointCircle`

Test if a point is inside a circle.

**Parameters**

- `point` (`Point`)
- `circle` (`Circle`)

[Source](https://github.com/studiometa/js-toolkit/blob/master/src/utils/collide/collidePointCircle.js)

**Usage**

```js
import { collidePointCircle } from '@studiometa/js-toolkit/utils/collide';

const point = {
  x: 25,
  y: 25,
};

const circle = {
  x: 40,
  y: 40,
  radius: 40,
};

collidePointCircle(point, circle); // true
```
