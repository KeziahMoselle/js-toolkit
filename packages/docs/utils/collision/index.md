# Collision

Collision utils can be used to test if two DOM elements are overlapping.

- [boundingRectToCircle](./boundingRectToCircle.html)
- [collideCircleCircle](./collideCircleCircle.html)
- [collideCircleRect](./collideCircleRect.html)
- [collidePointCircle](./collidePointCircle.html)
- [collidePointRect](./collidePointRect.html)
- [collideRectRect](./collideRectRect.html)

Inspired by http://www.jeffreythompson.org/collision-detection/.
