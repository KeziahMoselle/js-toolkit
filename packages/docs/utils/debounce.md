# `debounce`

Execute a function until it stops being called.

**Parameters**

- `fn` (`Function`): the function to execute
- `delay` (`Number`): the delay in milliseconds

[Source](https://github.com/studiometa/js-toolkit/blob/master/src/utils/debounce.js)

**Usage**

```js
import debounce from '@studiometa/js-toolkit/utils/debounce';

const debouncedFn = debounce(() => {
  console.log('Hello ðŸ‘‹');
}, 500);

debouncedFn(); // Hello ðŸ‘‹
```

::: tip
You can read [Debounce vs throttle](https://redd.one/blog/debounce-vs-throttle) if the difference between the `debounce` and `throttle` methods is not clear for you.
:::
