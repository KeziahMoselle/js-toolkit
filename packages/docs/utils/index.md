

# Utils


## `matrix`

This method will format a matrix CSS transform function.

**Parameters**

- `options.scaleX` (`Number`): the scale on the x axis
- `options.scaleY` (`Number`): the scale on the y axis
- `options.skewX` (`Number`): the skew on the x axis
- `options.skewY` (`Number`): the skew on the y axis
- `options.translateX` (`Number`): the translate on the x axis
- `options.translateY` (`Number`): the translate on the y axis

**Returns**

- `String`: a formatted CSS matrix transform

[Source](https://github.com/studiometa/js-toolkit/blob/master/src/utils/css/matrix.js)

**Usage**
```js
import matrix from '@studiometa/js-toolkit/utils/css/matrix';

document.body.style.transform = matrix({ scaleX: 0.5, scaleY: 0.5 });
// matrix(0.5, 0, 0, 0.5, 0, 0)
```

## `transition`

This method implements the same logic as the [`<transition>` component](https://vuejs.org/v2/api/#transition) from Vue to apply CSS based transition to a DOM element.

**Parameters**

- `element` (`HTMLElement`): the target DOM element
- `name` (`String|Object`): the name of the transition or an object of classes or styles to apply
- `endMode = 'remove'` (`String`): whether or not the transition keeps its final state or not, can be `remove` or `keep`, defaults to `remove`.

**Returns**

- `Promise`: a Promise

[Source](https://github.com/studiometa/js-toolkit/blob/master/src/utils/css/transition.js)

**Usage**
```js
import transition from '@studiometa/js-toolkit/utils/css/transition';

// Will apply and remove the following classes sequentially:
// - fade-from
// - fade-active
// - fade-to
transition(document.body, 'fade');

// Will apply and remove the following styles sequentially:
// - opacity: 0;
// - transition: opacity 1s ease-out
transition(document.body, {
  from: { opacity: 0 },
  active: { transition: 'opacity 1s ease-out' },
});

// Will apply and remove the following classes sequentially:
// - opacity-0
// - transition duration-500
// - opacity-50
transition(document.body, {
  from: 'opacity-0',
  active: 'transition duration-500',
  to: 'opacity-50',
});

// Will apply and remove the following classes sequentially:
// - opacity-0
// - transition duration-500
// And keep the latest class at the end:
// - opacity-50
transition(document.body, {
  from: 'opacity-0',
  active: 'transition duration-500',
  to: 'opacity-50',
}, 'keep');
```

## [Collide](./collide.md)
Collides utils is used to test if two DOM elements are overlapping.
[Go to the detailed documentation.](./collide.md)
