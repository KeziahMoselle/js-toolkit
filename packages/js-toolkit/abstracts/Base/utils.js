/**
 * @typedef {import('./index.js').default} Base
 */

/**
 * Test if an object has a method.
 *
 * @param  {Object}  obj  The object to test
 * @param  {String}  name The method's name
 * @return {Boolean}
 */
export function hasMethod(obj, name) {
  return typeof obj[name] === 'function';
}

/**
 * Call the given method while applying the given arguments.
 *
 * @param {Base}   instance The Base instance on which to trigger the method.
 * @param {String} method   The method to call.
 * @param {...any} args     The arguments to pass to the method.
 */
export function callMethod(instance, method, ...args) {
  instance.__debug('callMethod', method, ...args);

  // Prevent duplicate call of `mounted` and `destroyed`
  // methods based on the component status
  if (
    (method === 'destroyed' && !instance.$isMounted) ||
    (method === 'mounted' && instance.$isMounted)
  ) {
    instance.__debug('not', method, 'because the method has already been triggered once.');
    return instance;
  }

  instance.$emit(method, ...args);

  // We always emit an event, but we do not call the method if it does not exist
  if (!hasMethod(instance, method)) {
    return instance;
  }

  instance[method].call(instance, ...args);
  instance.__debug(method, instance, ...args);

  return instance;
}

/**
 * Get a list of elements based on the name of a component.
 *
 * @param {String} nameOrSelector
 *   The name or selector to used for this component.
 * @param {HTMLElement|Document} element
 *   The root element on which to query the selector, defaults to `document`.
 * @return {Array<HTMLElement>}
 *   A list of elements on which the component should be mounted.
 */
export function getComponentElements(nameOrSelector, element = document) {
  const selector = `[data-component="${nameOrSelector}"]`;
  let elements = [];

  try {
    elements = Array.from(element.querySelectorAll(selector));
    // eslint-disable-next-line no-empty
  } catch {}

  // If no child component found with the default selector, try a classic DOM selector
  if (elements.length === 0) {
    elements = Array.from(element.querySelectorAll(nameOrSelector));
  }

  return elements;
}
