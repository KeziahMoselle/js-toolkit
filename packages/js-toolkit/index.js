import Base from './abstracts/Base/index.js';
import defineComponent from './helpers/defineComponent.js';
import createBase from './helpers/createBase.js';

export { Base, createBase, defineComponent };

export default Base;
