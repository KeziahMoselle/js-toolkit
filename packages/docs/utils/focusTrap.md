# focusTrap

Trap the tab navigation inside a given element.

::: tip
To understand the "tab trap" usage, read [Using JavaScript to trap focus in an element](https://hiddedevries.nl/en/blog/2017-01-29-using-javascript-to-trap-focus-in-an-element).
:::

## Usage

```js
import { focusTrap } from '@studiometa/js-toolkit/utils';

const { trap, untrap } = focusTrap();

// Limit the tab navigation to focusable children of the document's body
document.addEventListener('keyup', (event) => {
  trap(document.body, event);
});

// Resume the trap and refocus the previously focused element
untrap();
```
