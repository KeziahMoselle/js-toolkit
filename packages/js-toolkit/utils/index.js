import debounce from './debounce.js';
import focusTrap from './focusTrap.js';
import keyCodes from './keyCodes.js';
import memoize from './memoize.js';
import nextFrame from './nextFrame.js';
import throttle from './throttle.js';
import { matrix, transition } from './css/index.js';
import {
  collidePointRect,
  collideRectRect,
  boundingRectToCircle,
  collideCircleRect,
  collideCircleCircle,
  collidePointCircle,
} from './collide/index.js';
import {
  objectToURLSearchParams,
  push as historyPush,
  replace as historyReplace,
} from './history.js';
import { clamp, clamp01, damp, inertiaFinalValue, lerp, map, round } from './math/index.js';
import {
  withLeadingCharacters,
  withLeadingSlash,
  withoutLeadingCharacters,
  withoutLeadingCharactersRecursive,
  withoutLeadingSlash,
  withoutTrailingCharacters,
  withoutTrailingCharactersRecursive,
  withoutTrailingSlash,
  withTrailingCharacters,
  withTrailingSlash,
} from './string/index.js';

// css
export { matrix, transition };

// collide
export {
  collidePointRect,
  collideRectRect,
  boundingRectToCircle,
  collideCircleRect,
  collideCircleCircle,
  collidePointCircle,
};

// history
export { objectToURLSearchParams, historyPush, historyReplace };

// math
export { clamp, clamp01, damp, inertiaFinalValue, lerp, map, round };

// string
export {
  withLeadingCharacters,
  withLeadingSlash,
  withoutLeadingCharacters,
  withoutLeadingCharactersRecursive,
  withoutLeadingSlash,
  withoutTrailingCharacters,
  withoutTrailingCharactersRecursive,
  withoutTrailingSlash,
  withTrailingCharacters,
  withTrailingSlash,
};

export { debounce, focusTrap, keyCodes, memoize, nextFrame, throttle };
