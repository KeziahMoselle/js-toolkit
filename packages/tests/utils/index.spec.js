import * as utils from '@studiometa/js-toolkit/utils';

describe('@studiometa/js-toolkit/utils exports', () => {
  it('should export all scripts', () => {
    expect(Object.keys(utils)).toEqual([
      'collide',
      'css',
      'debounce',
      'focusTrap',
      'history',
      'keyCodes',
      'math',
      'memoize',
      'nextFrame',
      'object',
      'string',
      'throttle',
    ]);
  });
});
