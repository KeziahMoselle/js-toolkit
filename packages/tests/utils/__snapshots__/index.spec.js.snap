// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@studiometa/js-toolkit/utils exports should export all scripts 1`] = `
Array [
  "boundingRectToCircle",
  "clamp",
  "clamp01",
  "collideCircleCircle",
  "collideCircleRect",
  "collidePointCircle",
  "collidePointRect",
  "collideRectRect",
  "damp",
  "debounce",
  "focusTrap",
  "historyPush",
  "historyReplace",
  "inertiaFinalValue",
  "keyCodes",
  "lerp",
  "map",
  "matrix",
  "memoize",
  "nextFrame",
  "objectToURLSearchParams",
  "round",
  "throttle",
  "transition",
  "withLeadingCharacters",
  "withLeadingSlash",
  "withTrailingCharacters",
  "withTrailingSlash",
  "withoutLeadingCharacters",
  "withoutLeadingCharactersRecursive",
  "withoutLeadingSlash",
  "withoutTrailingCharacters",
  "withoutTrailingCharactersRecursive",
  "withoutTrailingSlash",
]
`;
